### ISSUE TITLE
Applying fixes does not work

### ISSUE THREAD

**Issue by vkucera** [megalinter/megalinter#481]
**Describe the bug**

The only way of applying fixes that seems to work is the commit mode on pull request events from one branch to another in the same repository. All other cases fail for various reasons.

**To Reproduce**
* Create a new branch.
* Commit changes that require fixes.
* Push to the remote branch.
* Create a pull request to the main branch.
* See the log.
* Repeat for all combinations of `APPLY_FIXES_EVENT` and `APPLY_FIXES_MODE`.

**Expected behaviour**
Fixes are applied as commit or pull request (according to `APPLY_FIXES_MODE`) for push or pull request events (according to `APPLY_FIXES_EVENT`).

**Additional context**
* Applying of fixes on push events is blocked by the condition `github.event.pull_request.head.repo.full_name == github.repository` which can never be satisfied because `github.event.pull_request` exists only for pull request event.
* Applying of fixes in the pull request mode does not work because the repository is checked out in a detached head and the `peter-evans/create-pull-request` action complains with:
```
fatal: ref HEAD is not a symbolic ref
Error: When the repository is checked out on a commit instead of a branch, the 'base' input must be supplied.
```
I tried to run the workflow with the branch `HEAD` checked out using: `ref: ${{ github.event.pull_request.head.sha }}` in the `checkout` action but that makes the Mega-Linter runner fail with:
```
Warning: Unable to pull current branch <SHA>
cmdline: git diff --name-only master...<SHA>
stderr: 'fatal: Invalid symmetric difference expression master...<SHA>
```
where `<SHA>` is a commit SHA which is different from the branch `HEAD` at which the repository was initially checked out.
(Could it be the commit with the fixes?)

**How about pull requests between repositories?**
Besides fixing these issues it would also be very useful if the fixes could be applied in pull request events from a fork to an upstream repository, since that is the way pull requests are mostly used.

**I don't know if it is related, but for completeness**
For tests of pull requests from a fork to an upstream repository, this warning always appears:
```
Listing updated files in [/github/workspace] using git diff, then filter with:
Warning: Unable to pull current branch <SHA>
Git diff :
```
where `<SHA>` corresponds to the merge commit created by the checkout action.
This is then followed by this error for each linter:
```
[GitHub Status Reporter] Error posting Status for <CATEGORY>with <linter>: 403
GitHub API response: {"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/reference/repos#create-a-commit-status"}
```

**Comment by nvuillam** [comment]
Thanks for this feedback, I think you're the first to check other cases than Commit withing the same repo !

If you have proposition of fixes, please be my guest to make a PR
Else i'll try to check all that one of these week-ends...

**Comment by vkucera** [comment]
Allowing the commit mode for push events can be done simply by replacing
`github.event.pull_request.head.repo.full_name == github.repository`
with
`(github.event_name == 'push' || github.event.pull_request.head.repo.full_name == github.repository)`
That one is quite straightforward. But if you prefer, I can make a PR.

I don't have any further suggestions on the other errors, unfortunately.

**Comment by nvuillam** [comment]
@vkucera please make a PR, that will make you a contributor :)

**Comment by vkucera** [comment]
I followed the instructions at https://nvuillam.github.io/mega-linter/contributing/
but running `./build.sh` ends with:
```
Formatting markdown tables...
npx: installed 17 in 1.736s
Unexpected token ;
```
Do you have an idea where this error comes from?

**Comment by nvuillam** [comment]
Did you install https://www.npmjs.com/package/markdown-table-formatter ?

**Comment by nvuillam** [comment]
and do you have latest node.js version ?

**Comment by vkucera** [comment]
> Did you install https://www.npmjs.com/package/markdown-table-formatter ?

No, I didn't. The instructions mention installation of `markdown-table-formatter` only if the previous steps didn't work. In my case they did work, so I had no reason to do so.

> and do you have latest node.js version ?

No I don't. My version is v10.19.0 (available in Ubuntu 20.04). `node.js` is not mentioned anywhere in the instructions. If only specific versions are required, it should be mentioned.

Btw, the line `pip install -r requirements.dev.` seems to be missing the `txt` extension.

**Comment by vkucera** [comment]
not stale

**Comment by nvuillam** [comment]
Agreed... sorry my rent-paying job is very time consuming these days !

**Comment by dgokcin** [comment]
Hello



What I am trying to do is analyzing the newly changed files against the .editorconfig file at the root of my repository and if there are any broken properties, fixing them automatically(either via a commit or a pr, doesn't matter). 

When I create a feature branch, create an issue and create a PR, the editorconfig scan gives an error as expected and the steps about creating a pr and fixing things are skipped. I am adding screenshots and my workflow to describe the issue ina better way. Am I missing anything? 

Looking forward for your help @nvuillam 

Thanks

<img width="640" alt="Screen Shot 2021-08-10 at 16 23 45" src="https://user-images.githubusercontent.com/33603535/128874735-2460965b-5b1e-4b8f-bfda-0468272c975c.png">
<img width="303" alt="Screen Shot 2021-08-10 at 16 24 02" src="https://user-images.githubusercontent.com/33603535/128874776-86c4b792-4e19-4810-bc95-b7e20f838106.png">


```yaml
---
# Mega-Linter GitHub Action configuration file
# More info at https://nvuillam.github.io/mega-linter
name: Mega-Linter

on:
  # Trigger mega-linter at every push. Action will also be visible from Pull Requests to master
  #push: # Comment this line to trigger action only on pull-requests (not recommended if you don't pay for GH Actions)
  pull_request:
    branches: [master, develop]

env: # Comment env block if you do not want to apply fixes
  # Apply linter fixes configuration
  APPLY_FIXES: all # When active, APPLY_FIXES must also be defined as environment variable (in github/workflows/mega-linter.yml or other CI tool)
  APPLY_FIXES_EVENT: pull_request # Decide which event triggers application of fixes in a commit or a PR (pull_request, push, all)
  APPLY_FIXES_MODE: commit # If APPLY_FIXES is used, defines if the fixes are directly committed (commit) or posted in a PR (pull_request)

jobs:
  # Cancel duplicate jobs: https://github.com/fkirc/skip-duplicate-actions#option-3-cancellation-only
  cancel_duplicates:
    name: Cancel duplicate jobs
    runs-on: ubuntu-latest
    steps:
      - uses: fkirc/skip-duplicate-actions@master
        with:
          github_token: ${{ secrets.PAT || secrets.GITHUB_TOKEN }}

  build:
    name: Mega-Linter
    runs-on: ubuntu-latest
    steps:
      # Git Checkout
      - name: Checkout Code
        uses: actions/checkout@v2
        with:
          token: ${{ secrets.PAT || secrets.GITHUB_TOKEN }}
          fetch-depth: 0

      # Mega-Linter
      - name: Mega-Linter
        id: ml
        # You can override Mega-Linter flavor used to have faster performances
        # More info at https://nvuillam.github.io/mega-linter/flavors/
        uses: nvuillam/mega-linter@v4
        env:
          # All available variables are described in documentation
          # https://nvuillam.github.io/mega-linter/configuration/
          VALIDATE_ALL_CODEBASE: false
          DEFAULT_BRANCH: develop
          #VALIDATE_ALL_CODEBASE: ${{ github.event_name == 'push' && github.ref == 'refs/heads/master' }} # Validates all source when push on master, else just the git diff with master. Override with true if you always want to lint all sources
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          # ADD YOUR CUSTOM ENV VARIABLES HERE OR DEFINE THEM IN A FILE .mega-linter.yml AT THE ROOT OF YOUR REPOSITORY
          # DISABLE: COPYPASTE,SPELL # Uncomment to disable copy-paste and spell checks

      # Upload Mega-Linter artifacts
      - name: Archive production artifacts
        if: ${{ success() }} || ${{ failure() }}
        uses: actions/upload-artifact@v2
        with:
          name: Mega-Linter reports
          path: |
            report
            mega-linter.log

      # Create pull request if applicable (for now works only on PR from same repository, not from forks)
      - name: Create Pull Request with applied fixes
        id: cpr
        if: steps.ml.outputs.has_updated_sources == 1 && (env.APPLY_FIXES_EVENT == 'all' || env.APPLY_FIXES_EVENT == github.event_name) && env.APPLY_FIXES_MODE == 'pull_request' && (github.event_name == 'push' || github.event.pull_request.head.repo.full_name == github.repository)
        uses: peter-evans/create-pull-request@v3
        with:
          token: ${{ secrets.PAT || secrets.GITHUB_TOKEN }}
          commit-message: "[Mega-Linter] Apply linters automatic fixes"
          title: "[Mega-Linter] Apply linters automatic fixes"
          labels: bot
      - name: Create PR output
        if: steps.ml.outputs.has_updated_sources == 1 && (env.APPLY_FIXES_EVENT == 'all' || env.APPLY_FIXES_EVENT == github.event_name) && env.APPLY_FIXES_MODE == 'pull_request' && (github.event_name == 'push' || github.event.pull_request.head.repo.full_name == github.repository)
        run: |
          echo "Pull Request Number - ${{ steps.cpr.outputs.pull-request-number }}"
          echo "Pull Request URL - ${{ steps.cpr.outputs.pull-request-url }}"

      # Push new commit if applicable (for now works only on PR from same repository, not from forks)
      - name: Prepare commit
        if: steps.ml.outputs.has_updated_sources == 1 && (env.APPLY_FIXES_EVENT == 'all' || env.APPLY_FIXES_EVENT == github.event_name) && env.APPLY_FIXES_MODE == 'commit' && github.ref != 'refs/heads/master' && (github.event_name == 'push' || github.event.pull_request.head.repo.full_name == github.repository)
        run: sudo chown -Rc $UID .git/
      - name: Commit and push applied linter fixes
        if: steps.ml.outputs.has_updated_sources == 1 && (env.APPLY_FIXES_EVENT == 'all' || env.APPLY_FIXES_EVENT == github.event_name) && env.APPLY_FIXES_MODE == 'commit' && github.ref != 'refs/heads/master' && (github.event_name == 'push' || github.event.pull_request.head.repo.full_name == github.repository)
        uses: stefanzweifel/git-auto-commit-action@v4
        with:
          branch: ${{ github.event.pull_request.head.ref || github.head_ref || github.ref }}
          commit_message: "[Mega-Linter] Apply linters fixes"

```

**Comment by nvuillam** [comment]
@dgokcin to apply fixes with a new PR or commit, there must be no blocking lint error on the Mega-Linter step

If you want to apply fixes while there are still not auto-fixable lint errors, you can copy the content of "updated_files" in report artifact

**Comment by vkucera** [comment]
Hi @nvuillam , I am wondering whether there has been any progress on this. Especially the option to apply formatting as a PR would be very much appreciated.

**Comment by nvuillam** [comment]
Not yet, sorry

**Comment by Kurt-von-Laven** [comment]
See [#2125](https://github.com/oxsecurity/megalinter/issues/2125#issuecomment-1445286419).
